src/lib/js/filter.js

```diff
@@ -0,0 +1,149 @@
+import _ from 'lodash';
+
+export function filterByMultiple(data, filter_values, property) {
+  return data.filter((object) => {
+    const objectsProperty = object[property];
+    return filter_values.every((single) => {
+      if (single === 'global' && property === 'correlaidx') {
+        return objectsProperty.length === 0;
+      } else {
+        return objectsProperty
+          .map((entry) => entry.toLowerCase())
+          .includes(single.toLowerCase());
+      }
+    });
+  });
+}
+
+function filterDefinedBy(property, objects, value) {
+  return _.filter(objects, (object) => {
+    return object && object[property].toLowerCase() === value.toLowerCase();
+  });
+}
+
+export function filterStringSearch(searchTerm, searchOptions, objects) {
+  return _.filter(objects, (object) => {
+    let bool = false;
+    for (const item of searchOptions) {
+      if (item.multiple === true) {
+        for (const single of object[item.name]) {
+          if (single.toLowerCase().includes(searchTerm.toLowerCase())) {
+            bool = true;
+          }
+        }
+      } else {
+        if (
+          object &&
+          object[item.name].toLowerCase().includes(searchTerm.toLowerCase())
+        ) {
+          bool = true;
+        }
+      }
+    }
+
+    return bool;
+  });
+}
+
+export function filter(data, selects, searchTerm, searchOptions) {
+  let data_ = structuredClone(data);
+  if (searchTerm) {
+    data_ = filterStringSearch(searchTerm, searchOptions, data_);
+  }
+
+  for (const select of selects) {
+    if (select.value) {
+      if (select.multiple === true) {
+        const values = _.chain(select.value).flatMap('value').value();
+        data_ = filterByMultiple(data_, values, select.param);
+      } else {
+        data_ = filterDefinedBy(select.param, data_, select.value.value);
+      }
+    }
+  }
+  return data_;
+}
+
+export function setUrlParams(url, selects) {
+  const newUrl = new URL(url);
+  for (const select of selects) {
+    if (select.value) {
+      if (select.multiple === true) {
+        newUrl?.searchParams?.set(
+          select.param,
+          _.chain(select.value).flatMap('value').value(),
+        );
+      } else {
+        newUrl?.searchParams?.set(select.param, select.value.value);
+      }
+    } else {
+      newUrl?.searchParams?.delete(select.param);
```

Comment:
ists for the case that newUrl is undefined. If the url parameter passed to the setUrlParams function is an invalid URL or if the URL constructor fails, it would result in newUrl being undefined. but as the input url is taken from the actual page url this will probably not happen so ill remove ?